overlord/servicestate/quota_control.go

```diff
@@ -164,33 +198,56 @@ type QuotaGroupUpdate struct {
 // TODO: this should support more kinds of updates such as moving groups between
 // parents, removing sub-groups from their parents, and removing snaps from
 // the group.
-func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) error {
+func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) (*state.TaskSet, error) {
 	if err := quotaGroupsAvailable(st); err != nil {
-		return err
+		return nil, err
 	}
 
+	// TODO: conflict checking for other changes with this quota group
+
 	allGrps, err := AllQuotas(st)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	grp, ok := allGrps[name]
+	if !ok {
+		return nil, fmt.Errorf("group %q does not exist", name)
 	}
 
-	// TODO: switch to returning a taskset with the right handler instead of
-	// executing this directly
+	// check that the memory limit is not being decreased
+	if updateOpts.NewMemoryLimit != 0 {
+		// we disallow decreasing the memory limit because it is difficult to do
+		// so correctly with the current state of our code in
+		// EnsureSnapServices, see comment in ensureSnapServicesForGroup for
+		// full details
+		if updateOpts.NewMemoryLimit < grp.MemoryLimit {
+			return nil, fmt.Errorf("cannot decrease memory limit of existing quota-group, remove and re-create it to decrease the limit")
```

Comment:
AFAICT this case isn't covered either.

Reply:
I think I now pushed tests that cover most of these